pandas/core/frame.py

```diff
@@ -8942,6 +8942,17 @@ def aggregate(self, func=None, axis: Axis = 0, *args, **kwargs):
 
     agg = aggregate
 
+    @doc(NDFrame.any, **_shared_doc_kwargs)
+    def any(
+        self: DataFrame,
+        axis: Axis = 0,
+        bool_only: bool | None = None,
+        skipna: bool = True,
+        level: Level | None = None,
+        **kwargs,
+    ) -> DataFrame | Series:
```

Comment:
The following is not necessary, feel free to ignore it.

It would be great to add overloads (depending on the type of level). The issue is that `None` is `Hashable` and `Level = Union[Hashable, int]`. I think the best we could do is to give a precise type in the default case of `None` and return a Union otherwise.

Something like this might work (did not test it)
```py
    @overload
    def any(
        self,
        axis: Axis = ...,
        bool_only: bool | None = ...,
        skipna: bool = ...,
        *
        level: None = None,
        **kwargs,
    ) -> Series:
        ...

    @overload
    def any(
        self,
        axis: Axis = ...,
        bool_only: bool | None = ...,
        skipna: bool = ...,
        *
        level: None,
        **kwargs,
    ) -> Series:
        ...

    @overload
    def any(
        self,
        axis: Axis = ...,
        bool_only: bool | None = ...,
        skipna: bool = ...,
        *
        level: Level,
        **kwargs,
    ) -> DataFrame | Series:
        ...
```

You might want to test this with mypy (or pyright)
```py
reveal_type(pd.DataFrame().any())
reveal_type(pd.DataFrame().any(level=None))
reveal_type(pd.DataFrame().any(level="something"))
```

Reply:
@twoertwein the issue is that then
```
df.any(0, None, None, None)
```
would result in
```
t.py:4: error: No overload variant of "any" of "DataFrame" matches argument types "int", "None", "None", "None"  [call-overload]
t.py:4: note: Possible overload variants:
t.py:4: note:     def any(self, axis: Union[str, int] = ..., bool_only: Optional[bool] = ..., skipna: bool = ..., *level: None, **kwargs: Any) -> Series
t.py:4: note:     def any(self, axis: Union[str, int] = ..., bool_only: Optional[bool] = ..., skipna: bool = ..., *level: Union[Hashable, int], **kwargs: Any) -> Union[DataFrame, Series]
t.py:4: note: Revealed type is "Any"
Found 1 error in 1 file (checked 1 source file)
```

I think we're stuck between:
- waiting for arguments to become keyword-only
- adding 7 overload (!)
- assuming `any` will only be called with max 1 or 2 positional arguments

I think I'd have a preference for adding all overloads, and then removing most of them when the keyword-only deprecation takes places. Thoughts?

Reply:
> * waiting for arguments to become keyword-only

In other PRs where we also have a deprecation warning, we write the overloads as if the keywords are keyword-only. Basically, you get two warnings: at runtime and at type checking time. This allows us to move a bit quicker with typing.

Reply:
Ah, nice!

In that case, there's _a lot_ of redundant overloads that can be removed immediately

So yeah, happy to go with what you've suggested - @yadav-sachin want to make this change with the overloads?

Reply:
I think this hasn't been formalized or written down. imho this is the first PR I'm aware of #45610

Reply:
OK I'll ask on Wednesday, then we can come back to this

Reply:
In yesterdays' meeting, it was agreed to type functions as if the deprecation was already taking place

So @yadav-sachin , do you want to follow @twoertwein 's suggestion in https://github.com/pandas-dev/pandas/pull/46638#discussion_r842251078 ? (I think the first overload might be unnecessary)

Reply:
> I think the first overload might be unnecessary

I think the first overload is needed when the user doesn't provide level. I didn't test it with mypy&pyright, but I think the second overload could be skipped (already included in the first overload).

Reply:
> but I think the second overload could be skipped

🙌 yes, that looks like it (I also think it might not be valid syntax?)

Reply:
> I also think it might not be valid syntax?

Are you referring to keyword-only arguments without default values? That is possible if they are listed after the `*`.